### PR TITLE
feature(events): add original_timestamp

### PR DESCRIPTION
**Pull request by dkropachev**
Some events we pick up from the outside of sct.
These events have original time when they occur,
 let's store it and handle filter according to it

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/sct_events/file_logger.py [hunk 1]
```diff
@@ -74,7 +74,10 @@ def run(self) -> None:
                 self.write_event(event=event, tee=LOGGER.info)
 
     def write_event(self, event: SctEvent, tee: Optional[Callable[[str], Any]] = None) -> None:
-        message = f"{event.formatted_timestamp}: {str(event).strip()}"
+        if event.original_timestamp:
+            message = f"{event.formatted_timestamp} <{event.formatted_timestamp}>: {str(event).strip()}"
```
Comment (aleksbykov) [inline sdcm/sct_events/file_logger.py:78]: Should be here original_formatted_timestamp?

Comment (vponomaryov) [reply sdcm/sct_events/file_logger.py:78]: Agree, looks like typo.

File: sdcm/sct_events/events_device.py [hunk 2]
```diff
@@ -152,7 +152,7 @@ def outbound_events(self,
         with suppress_interrupt():
             for events_counter.value, obj in enumerate(self.inbound_events(stop_event=stop_event), start=1):
                 for filter_key, filter_obj in list(filters.items()):
-                    if filter_obj.expire_time and filter_obj.expire_time < obj.timestamp:
+                    if filter_obj.expire_time and filter_obj.expire_time + 600 < obj.effective_timestamp:
```
Comment (vponomaryov) [inline sdcm/sct_events/events_device.py:155]: Shouldn't we make value "600" configurable or more obvious to be changed? Like we do it on lines 33-41 in this module...

**Comment by dkropachev** [comment]
GCE failure not related to this PR

**Comment by dkropachev** [comment]
@vponomaryov , issue has been fixed.

**Review by vponomaryov — APPROVED** [review]

File: sdcm/sct_events/events_device.py (hunk 2 above)
Comment (dkropachev) [reply sdcm/sct_events/events_device.py]: fixed

File: sdcm/sct_events/base.py [hunk 3]
```diff
@@ -268,14 +269,28 @@ def add_subevent_type(cls,
...
+    @property
+    def formatted_timestamp(self) -> str:
+        return self._formatted_timestamp(self.timestamp)
+
+    @property
+    def formatted_original_timestamp(self) -> str:
+        return self._formatted_timestamp(self.original_timestamp)
+
+    @property
+    def effective_timestamp(self) -> float:
```
Comment (bentsi) [inline sdcm/sct_events/base.py:289]: explain the property (add docstring)

**Review by vponomaryov — DISMISSED** [review]

File: sdcm/sct_events/base.py [hunk 4]
```diff
@@ -149,14 +151,33 @@ def add_subevent_type(cls,
...
+    @property
+    def timestamp(self) -> float:
+        """
+        Some events are sourced from the outside of SCT, for such events we try to catch original time when
+          it was created and store it into source_timestamp
+        This property returns source_timestamp if it is present and timestamp if it is not
+        """
+        if self.source_timestamp:
+            return self.source_timestamp
+        return self.event_timestamp
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/sct_events/base.py]: thanks, fixed

File: sdcm/sct_events/events_device.py [hunk 5]
```diff
@@ -143,20 +145,21 @@ def inbound_events(self, stop_event: StopEvent) -> Generator[Any, None, None]:
...
             for events_counter.value, obj in enumerate(self.inbound_events(stop_event=stop_event), start=1):
-                for filter_key, filter_obj in list(filters.items()):
-                    if filter_obj.expire_time and filter_obj.expire_time < obj.timestamp:
-                        if (isinstance(obj, LogEvent) and getattr(filter_obj, "filter_node", None) == obj.node) or \
-                                isinstance(filter_obj, EventsFilter):
-                            LOGGER.debug("%s: delete filter with uuid=%s", self, filter_key)
+                if filters_gc_next_hit < time.perf_counter():
+                    # Run filter GC once in FILTERS_GC_PERIOD seconds
+                    for filter_key, filter_obj in list(filters.items()):
+                        if filter_obj.is_deceased():
```
Comment (enaydanov) [inline sdcm/sct_events/events_device.py:160]: What about moving changes for the filtering stuff into the separate commit?

**Review by enaydanov — CHANGES_REQUESTED** [review]

**Comment by enaydanov** [comment]
Need to resolve conflicts.

File: sdcm/sct_events/events_device.py (hunk 5 above)
Comment (dkropachev) [reply sdcm/sct_events/events_device.py:160]: done

File: sdcm/sct_events/base.py [hunk 6]
```diff
@@ -269,23 +288,44 @@ class SystemEvent(SctEvent, abstract=True):
...
+            self.deceased_time = value + FILTER_EVENT_DECAY_TIME
+
+    def is_deceased(self) -> bool:
+        """
+        When filter is expired we keep it in memory in case of getting events with
+          timestamp shifted back in time
+        This property is signaling that filter could be cleaned up
+        """
+        return self.deceased_time and time.time() >= self.deceased_time
+
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/sct_events/base.py]: done

**Comment by dkropachev** [comment]
> Need to resolve conflicts.

done

**Review by vponomaryov — APPROVED** [review]

**Review by amoskong — APPROVED** [review]
Looks good.

And it solved the unit_test fail of test_search_power_off.
ticket: https://trello.com/c/lzLYk4nQ/3881-fix-testsearchpoweroff

**Review by roydahan — APPROVED** [review]

**Comment by roydahan** [comment]
@bentsi let's prioritize please.

**Review by enaydanov — APPROVED** [review]
Let's see what will happen next :)

**Comment by enaydanov** [comment]
@dkropachev it can't be cherry-picked to 2021.1 branch cleanly. Please, prepare a PR for 2021.1 branch.

**Comment by bentsi** [comment]
let's not backport it

**Pull request merged**